### DISCUSSION TITLE
Tips on best setup...

### DISCUSSION THREAD

**Discussion by elltingle** [joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script#18] [Q&A] [answered]
Hi, and thank you for sharing this script. I am totally new to all this but your instructions look very helpful, even for a beginner :)

I haven't started using it yet as I am still working out the best way to set up my mockups and input/output folders. I would be grateful for advice on the best way to do this, please. Sorry this is such a long message.

I sell disc-bound notebooks and would like to produce several mockups for each cover design image:
- Different sizes (A5, A4, A6, etc)
- Different disc options (metal, plastic)
- Cover set only (no discs, just front and back cover)
- Different corners (square, 6mm round, 1cm round)
I currently have one big mockup psd file which has layers for the discs, duplicated Smart Objects for the different cover sizes, and clipping masks with folders for the different corner options. I only have one SO per mockup, just different instances of it.

I can create any mockups I need by replacing the SO contents with a new cover design image, then turning the different layers and clipping masks on and off and exporting the different versions. It is quite tedious (!) so I would like to automate it as a batch process.

As I am creating a new mockup to work with this script anyway, I would like to make the process as efficient as possible. What would be the best way to design it (out of the options below - or another way)?

(1) Keep it as one psd file, and modify the script so it turns the different disc layers and clipping mask layers on and off, and exports the file with different names in between. So, fully automate my current process. I don't always need all the variations so it would be handy to be able to toggle them on and off.

(2) Keep one psd file and just turn the layers and masks on manually for each variation,  then run the script to go through the input files. 

(3) Have a separate folder for each variation. The psd file in each folder would  only have the relevant layers and masks active and I would run your script for each variation (e.g. A5, Metal discs, 1cm corners). They would all use the same input and output folders. I assume if I use the $ options for the file paths and keep the psd file name and target name the same then the script wouldn't change, and I could just edit it once then duplicate the folder and tweak the psd layers. 

Also, is it possible to name the output files after the input files but with a different extension for each variation?

Any advice would be very much appreciated.

Thanks!

**Comment by joonaspaakko** [comment, marked as answer]
Yea, number 1 should be possible. 

You can [add multiple mockups](https://joonaspaakko.gitbook.io/batch-mockup-smart-object-replacement-jsx-photosho/anatomy-of-the-settings-script#using-multiple-mockups) in one batch process and every single one can use the same mockup .psd file, if that is your want. In that example I'm basically feeding every mockup the same output options using a variable, but you can give each mockup their own custom output options, which in this case should be necessary in order to avoid overwriting the output files as they would all share the same output location by default and presumably same input filenames too.

> In the small code snippet below you can see that instead of `output: outputOpts`, I'm directly feeding it the object `output: { ... }`, which can then be customized for each mockup.
```js
// output: outputOpts,
output: {
  format: 'jpg', 
  path: '$/Output/', // You can add as many nested folders as you want: '$/Output/black-spine/color-red'
  // If you wanted some or all files in the same output folder, you could differentiate by customizing the filename:
  filename: '@input', // You can add static text and only the keywords are replaced: 'black-spine--color-red--@input'
},
```
> Both (`path`, `filename`) properties can contain spaces, it's just a habit of mine to use dash...

The script has options [`showLayers`](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/blob/0ed94149a8ff4c6c585a6770015f352afd9c6271/Examples/example-2/Batch%20replace%20-%20example%202.jsx#L34-L35), and [`hideLayers `](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/blob/0ed94149a8ff4c6c585a6770015f352afd9c6271/Examples/example-2/Batch%20replace%20-%20example%202.jsx#L34-L35), though the batch process always starts from the initial state of each mockup psd, so it might be easiest for you to save the mockup with all layers hidden that change from mockup to mockup and then just use the `showLayers` option to pick whatever layers you want to show in each mockup output.

While you can fairly easily comment out lines or change the values, you can also have multiple setting scripts for different types of outputs. 

> You could even combine settings scripts by including one inside another with `#include "settings-1.js"`. Or by making an `Action` in PS that triggers multiple settings scripts at once.

----
----

I can imagine your settings file might end up looking a bit crowded due to repetition... So you could add helper functions to simplify it. Something like this (I didn't test it at all):

```js

#include "script/Batch Mockup Smart Object Replacement.jsx" 

function addMockup( config ) {
  
  var options = {
    output: {
      path: './Output' + (config.path || ''), // Outputs next to the mockup psd
      format: config.format || 'jpg', // 'jpg', 'png', 'tif', 'psd', 'pdf'
      filename: config.filename ? config.filename + ' @input' : '@mockup - @input', 
    }, 
    showLayers: config.showLayers || [],
    mockupPath: '$/assets/example-mockup.psd', 
    smartObjects: [
      
      // Smart object #1
      {
        target: '@my-smart-object',
        input:  config.input || '$/input', 
        inputNested: true,
        align: 'center center',
        resize: 'fill', 
      },
      
    ]
  };
  
  return options;
  
};

mockups([
  
  addMockup({
    output: '/grey',
    filename: 'grey',
    showLayers: ['grey-stuff'],
    input: ['$/input'],
  }),
  addMockup({
    output: '/red',
    showLayers: ['red-stuff'],
    filename: 'red',
    input: ['$/input'],
  }),
  addMockup({
    output: '/green',
    showLayers: ['green-stuff'],
    filename: 'green',
    input: ['$/input'],
  }),
  
]);
```

**Reply by elltingle** [reply]
That's brilliant, thank you!
